File: ios/Core/CacheIndex.swift
import Foundation

// Pure LRU bookkeeping. Mirrored in android/.../core/CacheIndex.kt.

struct CacheEntry: Codable, Equatable {
  var bytes: Int64
  var lastAccess: Double
  /** Total length when known, so partial and complete can be told apart. */
  var totalBytes: Int64?
}

struct CacheIndex: Codable {
  private(set) var entries: [String: CacheEntry] = [:]
  var budget: Int64

  init(budget: Int64) { self.budget = budget }

  var totalBytes: Int64 { entries.values.reduce(0) { $0 + $1.bytes } }

  mutating func record(_ key: String, bytes: Int64, totalBytes: Int64? = nil, now: Double) {
    var entry = entries[key] ?? CacheEntry(bytes: 0, lastAccess: now, totalBytes: nil)
    entry.bytes = bytes
    entry.lastAccess = now
    if let totalBytes { entry.totalBytes = totalBytes }
    entries[key] = entry
  }

  mutating func touch(_ key: String, now: Double) {
    entries[key]?.lastAccess = now
  }

  mutating func remove(_ key: String) { entries.removeValue(forKey: key) }
  mutating func removeAll() { entries.removeAll() }

  func status(_ key: String) -> (state: String, bytes: Int64) {
    guard let e = entries[key], e.bytes > 0 else { return ("none", 0) }
    if let total = e.totalBytes, e.bytes >= total { return ("complete", e.bytes) }
    return ("partial", e.bytes)
  }

  /**
   Keys to delete, oldest first, to get back under 90% of the budget once it's exceeded.
   Pinned keys (current and next items) are never evicted.
   */
  func evictions(pinned: Set<String>) -> [String] {
    var total = totalBytes
    guard total > budget else { return [] }
    let target = budget / 10 * 9
    var out: [String] = []
    for (key, entry) in entries.sorted(by: { $0.value.lastAccess < $1.value.lastAccess }) {
      if total <= target { break }
      if pinned.contains(key) { continue }
      out.append(key)
      total -= entry.bytes
    }
    return out
  }
}

File: android/src/main/java/expo/modules/infinitemedia/core/CacheIndex.kt
package expo.modules.infinitemedia.core

// Pure LRU bookkeeping. Mirrors ios/Core/CacheIndex.swift.

data class CacheEntry(var bytes: Long, var lastAccess: Double, var totalBytes: Long? = null)

class CacheIndex(var budget: Long) {
  val entries = LinkedHashMap<String, CacheEntry>()

  val totalBytes: Long get() = entries.values.sumOf { it.bytes }

  fun record(key: String, bytes: Long, totalBytes: Long? = null, now: Double) {
    val entry = entries.getOrPut(key) { CacheEntry(0, now) }
    entry.bytes = bytes
    entry.lastAccess = now
    if (totalBytes != null) entry.totalBytes = totalBytes
  }

  fun touch(key: String, now: Double) {
    entries[key]?.lastAccess = now
  }

  fun remove(key: String) {
    entries.remove(key)
  }

  fun removeAll() = entries.clear()

  /** "none" | "partial" | "complete" to bytes. */
  fun status(key: String): Pair<String, Long> {
    val e = entries[key] ?: return "none" to 0L
    if (e.bytes <= 0) return "none" to 0L
    val total = e.totalBytes
    return if (total != null && e.bytes >= total) "complete" to e.bytes else "partial" to e.bytes
  }

  /** Keys to delete, oldest first, to get back under 90% once over budget. Pinned keys stay. */
  fun evictions(pinned: Set<String>): List<String> {
    var total = totalBytes
    if (total <= budget) return emptyList()
    val target = budget / 10 * 9
    val out = mutableListOf<String>()
    for ((key, entry) in entries.entries.sortedBy { it.value.lastAccess }) {
      if (total <= target) break
      if (key in pinned) continue
      out.add(key)
      total -= entry.bytes
    }
    return out
  }
}
